cmd/dra-example-kubeletplugin/driver.go

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2023 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	drapbv1 "k8s.io/kubelet/pkg/apis/dra/v1alpha2"
+
+	nascrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1"
+	nasclient "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1/client"
+)
+
+type driver struct {
+	nascrd    *nascrd.NodeAllocationState
+	nasclient *nasclient.Client
+	state     *DeviceState
+}
+
+func NewDriver(config *Config) (*driver, error) {
+	var d *driver
+	client := nasclient.New(config.nascrd, config.exampleclient.NasV1alpha1())
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := client.GetOrCreate()
+		if err != nil {
+			return err
+		}
+
+		err = client.UpdateStatus(nascrd.NodeAllocationStateStatusNotReady)
+		if err != nil {
+			return err
+		}
+
+		state, err := NewDeviceState(config)
+		if err != nil {
+			return err
+		}
+
+		err = client.Update(state.GetUpdatedSpec(&config.nascrd.Spec))
+		if err != nil {
+			return err
+		}
+
+		err = client.UpdateStatus(nascrd.NodeAllocationStateStatusReady)
+		if err != nil {
+			return err
+		}
+
+		d = &driver{
+			nascrd:    config.nascrd,
+			nasclient: client,
+			state:     state,
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+func (d *driver) Shutdown() error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := d.nasclient.Get()
+		if err != nil {
+			return err
+		}
+		return d.nasclient.UpdateStatus(nascrd.NodeAllocationStateStatusNotReady)
+	})
+}
+
+func (d *driver) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
```

Comment:
Should we do contextual logging in this repo?

Can also be added later.

Reply:
I'm indifferent. Let's do it as a follow-up.